File: 174.dungeon-game.js
/*
 * @lc app=leetcode id=174 lang=javascript
 *
 * [174] Dungeon Game
 */

// @lc code=start
/**
 * @param {number[][]} dungeon
 * @return {number}
 */
var calculateMinimumHP = function(dungeon) {
  // Bottom-up DP approach
  // Calculate minimum HP at every iteration
  // O(m * n) time
  // O(m * n) space
  let m = dungeon.length, n = dungeon[0].length
  let dp = new Array(m + 1).fill(new Array(n + 1).fill(Number.MAX_SAFE_INTEGER))

  for (let i = m -1; i >= 0; i--) {
    for (let j = n -1; j >= 0; j--) {
      if (i === m-1 && j === n -1) {
        // bottom right cell (princess)
        dp[i][j] = Math.min(0, dungeon[i][j])
      } else if (i === m-1) {
        // last row
        dp[i][j] = Math.min(0, dungeon[i][j] + dp[i][j+1])
      } else if (j === n-1) {
        // last col
        dp[i][j] = Math.min(0, dungeon[i][j] + dp[i+1][j])
      } else {
        dp[i][j] = Math.min(0, dungeon[i][j] + Math.max(dp[i][j+1], dp[i+1][j]))
      }
    }
  }

  return Math.abs(dp[0][0]) + 1
};
// @lc code=end

File: cpp/174.dungeon-game.cpp
/*
 * @lc app=leetcode id=174 lang=cpp
 *
 * [174] Dungeon Game
 */

// @lc code=start
class Solution
{
public:
    int calculateMinimumHP(vector<vector<int>> &dungeon)
    {
        // bottom-up DP approach
        // calculate minimum HP at every iteration
        // O(m*n) time O(m*n) space
        int m = dungeon.size(), n = dungeon[0].size();
        vector<vector<int>> dpDungeon;
        for (int i = 0; i < dungeon.size(); i++)
        {
            vector<int> newRow;
            for (int j = 0; j < dungeon[i].size(); j++)
            {
                newRow.push_back(INT_MAX);
            }
            dpDungeon.push_back(newRow);
        }

        for (int i = m - 1; i >= 0; i--)
        {
            for (int j = n - 1; j >= 0; j--)
            {
                if ((i == m - 1) && (j == n - 1))
                {
                    // bottom right cell (princess)
                    dpDungeon[i][j] = min(0, dungeon[i][j]);
                }
                else if (i == m - 1)
                {
                    // last row
                    dpDungeon[i][j] = min(0, dungeon[i][j] + dpDungeon[i][j + 1]);
                }
                else if (j == n - 1)
                {
                    // last col
                    dpDungeon[i][j] = min(0, dungeon[i][j] + dpDungeon[i + 1][j]);
                }
                else
                {
                    dpDungeon[i][j] = min(0, dungeon[i][j] + max(dpDungeon[i][j + 1], dpDungeon[i + 1][j]));
                }
            }
        }

        return abs(dpDungeon[0][0]) + 1;
    }
};
// @lc code=end
